components/provider/src/lib.rs

```diff
@@ -89,7 +89,155 @@ pub mod prelude {
     pub use crate::data_provider::DataResponse;
     pub use crate::data_provider::DataResponseBuilder;
     pub use crate::error::Error as DataError;
+
+    pub use crate::v2::DataProviderV2;
+    pub use crate::v2::DataReceiver;
+    pub use crate::v2::DataReceiverImpl;
+    pub use crate::v2::DataResponseV2;
 }
 
 // Also include the same symbols at the top level for selective inclusion
 pub use prelude::*;
+
+pub mod v2 {
+    use crate::error::Error;
+    use crate::prelude::*;
+    use downcast_rs::Downcast;
+    use icu_locid::LanguageIdentifier;
+    use std::any::Any;
+    use std::any::TypeId;
+    use std::borrow::Borrow;
+    use std::borrow::BorrowMut;
+    use std::borrow::Cow;
+    use std::fmt::Debug;
+
+    pub trait DataReceiver<'d, 'de> {
+        fn set_to(
```

Comment:
Can we also split this into a generic and non-generic version? The generic `set_to` version will have to have a `where Self: Sized` so that you can still make a trait object.

This lets people write receivers that are tailored to the data being received. Serde is _really fast_ if you use generics, and that can be desired at times.

Reply:
This function uses erased_serde.  I don't have a compete mental model of how that works, but on one side of `set_to` (the concrete implementation), you have the struct deserialization code, and on the other side (the call site in the DataProvider), you have the JSON/Bincode deserialization code.  I assume they talk to each other via dynamic dispatch.  So does that mean erased_serde is slower than plain serde?  I haven't taken measurements, but I haven't read anything in particular that would suggest that is the case.

Reply:
I will follow up here: https://github.com/unicode-org/icu4x/issues/422